### PR TITLE
HTTPCLIENT-2277: RFC 9111 conformance improvements / bug fixes

### PR DESCRIPTION
**Pull request by ok2c**
* Aligns the behavior of the caching protocol handlers with the specification requirements (sections 3 and 4 RFC 9111)
* Improved normalization of variant header elements
* Bug fixes

### CODE REVIEW

**Comment by ok2c** [comment]
@arturobernalg This PR is still work in progress. Do take a look at the changes in `ResponseCachingPolicy` and `CachedResponseSuitabilityChecker` and let me know if find anything disagreeable.

**Comment by ok2c** [comment]
@arturobernalg Please review.

**Comment by arturobernalg** [comment]
> @arturobernalg Please review.

I haven't forgotten about this PR.

**Review by arturobernalg — APPROVED** [review]
@ok2c  this change-set LGTM

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheKeyGenerator.java [hunk 1]
```diff
@@ -201,6 +209,61 @@ public static List<String> variantNames(final MessageHeaders message) {
...
+            }
+        } else {
+            normalizeElements(message.headerIterator(headerName), consumer);
+        }
+    }
+
+    @Internal
+    public static void normalizeElements(final Iterator<Header> iterator, final Consumer<String> consumer) {
+        final List<HeaderElement> elements = new ArrayList<>();
+        final Iterator<HeaderElement> it = new BasicHeaderElementIterator(iterator);
```
[the remark this answers is not in the document]
Comment (ok2c) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheKeyGenerator.java:230]: @arturobernalg Agreed. Corrected. Please take another look.

**Review by arturobernalg — APPROVED** [review]

File: httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheKeyGenerator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (arturobernalg) [reply httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/CacheKeyGenerator.java:230]: > @arturobernalg Agreed. Corrected. Please take another look.

Perfect. LGTM

**Pull request merged**